src/rabin.js

```diff
@@ -0,0 +1,193 @@
+from: https://github.com/scrypt-sv/rabin/blob/master/rabin.py
+*/
+const { toBigIntLE } = require('bigint-buffer');
+const { checkIfValidHexString,
+    decimalToHexString,
+    hexStringToBigInt,
+    bigIntAbsoluteValue } = require('./utils');
+let crypto;
+try {
+    crypto = require('crypto');
+} catch (err) {
+    throw('Error: crypto support is disabled!');
+}
+    
+function greatestCommonDivisor(a, b) {
+    if ((typeof a !== 'bigint') || (typeof b !== 'bigint'))
+        throw "Error: Incorrect argument(s) to greatestCommonDivisor()";
+
+    a = bigIntAbsoluteValue(a);
+    b = bigIntAbsoluteValue(b);
+    if (b > a) {
+        let t = a;
+        a = b;
+        b = t;
+    }
+    while (b > 0) {
+        let t = b;
+        b = a % b;
+        a = t;
+    }
+    return a;
+}
+
+// Calculates: base^exponent % modulus
+function powerMod(base, exponent, modulus) {
+    if (modulus === 1n) return 0n;
+    let result = 1n;
+    base = base % modulus;
+    while (exponent > 0n) {
+        if (exponent % 2n === 1n) //odd number
+            result = (result * base) % modulus;
+        exponent = exponent >> 1n; //divide by 2
+        base = (base * base) % modulus;
+    }
+    return result;
+}
+
+function rabinHashBytes(bytes) {
+    hBytes = crypto.createHash('sha256').update(bytes).digest();
+    let idx = hBytes.byteLength / 2;
+    let hl = crypto.createHash('sha256').update(hBytes.slice(0, idx)).digest();
+    let hr = crypto.createHash('sha256').update(hBytes.slice(idx, hBytes.byteLength)).digest();
+    return toBigIntLE(Buffer.concat([hl, hr]));
+}
+
+function calculateNextPrime(p) {
+    smallPrimesProduct = 3n * 5n * 7n * 11n * 13n * 17n * 19n * 23n * 29n;
+    while (greatestCommonDivisor(p, smallPrimesProduct) != 1) {
+        p = p + 4n;
+    }
+    if (powerMod(2n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    if (powerMod(3n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    if (powerMod(5n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    if (powerMod(17n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    return p;
+}
+
+function getPrimeNumber(p) {
+    while (p % 4n != 3n) {
+        p = p + 1n;
+    }
+    return calculateNextPrime(p);
+}
+
+const paddingBuffer = Buffer.from('00', 'hex');
+
+function root(dataBuffer, p, q, nRabin) {
+    let sig, x, paddingByteCount = 0;
+    while (true) {
+        let h1 = rabinHashBytes(dataBuffer);
+        x = rabinHashBytes(dataBuffer) % nRabin;
+        sig = powerMod(p, q - 2n, q) * p * powerMod(x, (q + 1n) / 4n, q);
+        sig = (powerMod(q, p - 2n, p) * q * powerMod(x, (p + 1n) / 4n, p) + sig) % (nRabin);
+        if (((sig * sig) % nRabin) === x) {
+            break;
+        }
+        dataBuffer = Buffer.concat([dataBuffer, paddingBuffer]);
+        paddingByteCount++;
+    }
+    return {
+        "signature": decimalToHexString(sig),
```

Comment:
We can also return signature in bigint as keys, so we don't need to convert back & forth. Users can convert to whatever format they want, like hex.